38-allowlist-management.md

```diff
@@ -0,0 +1,162 @@
+* Performance: Since the ima measurements could be very large, it has impact on query
+  efficiency when the allowlist does not participate in the agent processing, e.g., we
+  don't need to load the allowlist for updating agent state.
+
+
+## Motivation
+
+### Goals
+
+The goal is to provide allowlist management API and CLI for allowlist management,
+establish relationship between an allowlist and an agent. The allowlist used by agent
+can be changed when the host of agent is upgraded and the old allowlist doesn't apply.
+
+### Non-Goals
+
+Import allowlist from agent.
+
+## Proposal
+
+The proposal is to:
+
+* Create a new table named ``allowlists`` to hold allowlist data that original hosted
+  in the ``verifiermain`` table, namely: ``tpm_policy``, ``vtpm_policy``, ``and allowlist``.
+* Add a foreign key named ``allowlist_id`` in the ``verifiermain`` table to reference the
+  allowlist data that will be used for the agent.
+* Implements ``/v2/allowlists/`` API for Verifier to support CRD operations. Update an
+  allowlist is not included in the initial proposal but can be implemented if necessary.
+* Enhance ``tenant`` CLI to manage allowlists, when adding agent to the Verifier, user
+  specifies an existing ``allowlist`` name instead of allowlist data to complete the
+  registration.
+
+
+### User Stories (optional)
+
+
+#### Story 1
+
+A user can reuse allowlist by:
+
+* Create an allowlist record by posting policies to the Verifier.
+* Add an agent to the Verifier by providing Agent ID, IP Address and the allowlist name
+  the agent will use.
+* Add another agent running under the same environment to the Verifier with the same allowlist.
+
+#### Story 2
+
+A user can update agent with new allowlist by:
+
+* Create an allowlist record by posting new policies to the Verifier.
+* Update an agent to the Verifier by updating the allowlist name.
+* Agent will use new allowlist to verify system integrety.
+
+### Notes/Constraints/Caveats (optional)
+
+None
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+DB Changes:
+
+The ``allowlists`` table will be created as:
+
+```
+op.create_table('allowlists',
+                sa.Column('id', sa.Integer(), nullable=False),
+                sa.Column('name', sa.String(255), nullable=False),
+                sa.Column('tpm_policy', sa.Text(), nullable=True),
+                sa.Column('vtpm_policy', sa.Text(), nullable=True),
+                sa.Column('ima_policy', sa.Text(length=429400000), nullable=True),
+                sa.PrimaryKeyConstraint('id'),
+                sa.UniqueConstraint('name', name='uniq_allowlists0name'),
+                mysql_engine='InnoDB',
+                mysql_charset='UTF8'
+                )
+```
+
+The ``verifiermain.allowlist`` column will be renamed to ``ima_policy`` in the
+``allowlists`` which provides a better presentation.
+
+The ``verifiermain`` will have a new column as foreign key to the ``allowlists`` table:
+
+```
```

Comment:
Minor nit: Same comment as above about specifying the language in the code block.